torchaudio/prototype/functional/_ray_tracing.py

```diff
@@ -0,0 +1,156 @@
+            representing absorption coefficients of ``"west"``, ``"east"``, ``"south"``, and
+            ``"north"`` walls, respectively.
+            Or the shape must be `(6,)` if the room is a 3D room, representing absorption coefficients
+            of ``"west"``, ``"east"``, ``"south"``, ``"north"``, ``"floor"``, and ``"ceiling"``, respectively.
+            If ``e_absorption`` is a 2D Tensor, the shape must be `(num_bands, 4)` if the room is a 2D room,
+            or `(num_bands, 6)` if the room is a 3D room. ``num_bands`` is the number of frequency bands (usually 7),
+            but you can choose other values.
+        scattering(float or torch.Tensor, optional): The scattering coefficients of wall materials.
+            (Default: ``0.0``).  The shape and type of this parameter is the same as for ``e_absorption``.
+        mic_radius(float, optional): The radius of the microphone in meters. (Default: 0.5)
+        sound_speed (float, optional): The speed of sound in meters per second. (Default: ``343.0``)
+        energy_thres (float, optional): The energy level below which we stop tracing a ray. (Default: ``1e-7``).
+            The initial enery of each ray is ``2 / num_rays``.
+        time_thres (float, optional):  The maximal duration (in seconds) for which rays are traced. (Defaut: 10.0)
+        hist_bin_size (float, optional): The size (in seconds) of each bin in the output histogram. (Default: 0.004)
+    Returns:
+        (torch.Tensor): The 3D histogram(s) where the energy of the traced ray is recorded. Each bin corresponds
+            to a given time slot. The shape is `(channel, num_bands, num_bins)`
+            where ``num_bins = ceil(time_thres / hist_bin_size)``. If both ``e_absorption`` and ``scattering``
+            are floats, then ``num_bands == 1``.
+    """
+    D = room.shape[0]
+
+    if mic_array.ndim == 1:
+        mic_array = mic_array[None, :]
+    if mic_array.ndim != 2:
+        raise ValueError(
+            f"mic_array must be 1D tensor of shape D, or 2D tensor of shape (num_mics, D) "
+            f"where D is 2 or 3. Got shape = {mic_array.shape}."
+        )
+    if room.dtype not in (torch.float32, torch.float64):
+        raise ValueError(f"room must be of float32 or float64 dtype, got {room.dtype} instead.")
+    if not (room.dtype == source.dtype == mic_array.dtype):
+        raise ValueError(
+            "dtype of room, source and mic_array must be the same. "
+            f"Got {room.dtype}, {source.dtype}, and {mic_array.dtype}"
+        )
+    if not (D == source.shape[0] == mic_array.shape[1]):
+        raise ValueError(
+            "Room dimension D must match with source and mic_array. "
+            f"Got {D}, {source.shape[0]}, and {mic_array.shape[1]}"
+        )
+    if time_thres < hist_bin_size:
+        raise ValueError(f"time_thres={time_thres} must be greater than hist_bin_size={hist_bin_size}.")
+
+    num_walls = 4 if D == 2 else 6
```

Comment:
perhaps we can move this into `_validate_absorption_scattering` since the result is used only there